_source/_posts/2021-12-06-spring-data-elasticsearch.md

```diff
@@ -0,0 +1,349 @@
+
+```shell
+mkdir spring-data-elasticsearch
+cd spring-data-elasticsearch
+https -d https://raw.githubusercontent.com/indiepopart/spring-data-elasticsearch/5380a940a0364176a5d56349d3471f60a0834ed7/blog-reactive-ms.jdl
+jhipster jdl blog-reactive-ms.jdl
+```
+
+{% include setup/cli.md type="jhipster" %}
+
+Follow the process above for creating 2 applications in Okta, the gateway application, and the blog application.
+
+```shell
+cd gateway
+okta apps create jhipster
+```
+
+Make sure to set up http://localhost:8081/login/oauth2/code/oidc with port 8081 as Redirect URI for blog microservice.
+
+```shell
+cd blog
+okta apps create jhipster
+```
+
+Find the `blog` credentials in the files `blog/.okta.env` and  `gateway/.okta.env`.
+
+The JHipster Registry is also a Spring Cloud Config server, and by default, it is configured with the profiles `dev` and `native`, which means the configuration will be provided from the location `docker-compose/central-server/config`. Let's add some files so it can provide the OAuth2 configuration to the microservices.
+Add the file `docker-compose/central-server-config/blog-prod.yml` with the following content:
+
+```yml
+spring:
+  security:
+    oauth2:
+      client:
+        provider:
+          oidc:
+            issuer-uri: https://{yourOktaDomain}/oauth2/default
+        registration:
+          oidc:
+            client-id: {blogClientId}
+            client-secret: {blogClientSecret}
+```
+
+Add also the file `docker-compose/central-server-config/gateway-prod.yml` with the following content:
+
+```yml
+spring:
+  security:
+    oauth2:
+      client:
+        provider:
+          oidc:
+            issuer-uri: https://{yourOktaDomain}/oauth2/default
+        registration:
+          oidc:
+            client-id: {gatewayClientId}
+            client-secret: {gatewayClientSecret}
+```
+
+For the experimentation in this tutorial, configure the Kibana interface for Elasticsearch, which will allow visualizations of the Elasticsearch data.
+Edit `docker-compose/docker-compose.yml` and add the Kibana service like this:
+
+```yml
+services:
+  blog-kibana:
+     image: docker.elastic.co/kibana/kibana:7.13.3
+     ports:
+       - 5601:5601
+     environment:
+       ELASTICSEARCH_URL: http://blog-elasticsearch:9200
+       ELASTICSEARCH_HOSTS: '["http://blog-elasticsearch:9200"]'
+```
+
+Also add the following additional configuration to the elasticsearch service, to expose the ports to the local host for sending APIs requests:
+
+```yml
+services:
+  blog-elasticsearch:
+    ...
+    ports:
+      - 9200:9200
+```
+
+Create the applications container image:
+
+```shell
+cd blog
+./mvnw -DskipTests -ntp -Pprod verify jib:dockerBuild
+```
+
+```shell
+cd gateway
+./mvnw -DskipTests -ntp -Pprod verify jib:dockerBuild
+```
+
+Run the architecture with Docker Compose:
+
+```shell
+cd docker-compose
+docker compose up
```

Comment:
This doesn't work for me. When I log in to the JHipster Registry, I see the blog service with a `OUT_OF_SERVICE` status. The Docker dashboard shows it running, but it doesn't work if I navigate to the blog entities via `http://localhost:8080`. I am on a slow internet connection so that might have something to do with it.

Reply:
I am not sure why it happens, after I changed the code to use a single Okta client app, I could replicate the error. Right now, if I set MANAGEMENT_ENDPOINT_HEALTH_PROBES_ENABLED=false for the blog service, it does not happen any more. Seems related to:
https://github.com/spring-cloud/spring-cloud-netflix/issues/3941

"The microservices not working first send a OUT_OF_SERIVCE. So far so good.Problem is, Eureka discards all of the following Heartbeats with state UP and the service is in OUT_OF_SERVICE forever."

https://spring.io/blog/2020/03/25/liveness-and-readiness-probes-with-spring-boot

Do I just add this config to the blog post and code?

Reply:
I would add it as a potential issue that can happen in the blog post. It's possible we could add some waits in the `docker-compose.yml` file that starts the services in a certain order. I'm not sure. I can QA again when you're ready. I'm back home in the US and have faster internet speeds now.

Reply:
I tried depends_on and it did not work. Maybe because it does not wait for the service to be ready, only to be started. Will add the potential issue note.